Expose UDP message on /metrics endpoint

fmejia97 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:

This PR:
1. Adds a UDP collector that listens to UDP messages from `monitor.lua` and exposes them on `/metrics` endpoint using prometheus client.
2. Removes VTS collector. 

**Why do we need this PR?** 

With the dynamic mode introduced a couple of releases ago, the load balancing is handled in LUA. This means there is no upstream section in the generated nginx.conf, there is only one used to switch to LUA. In this scenario, the VTS module is useless because the stats only contains information about this upstream without any context about the traffic. Additionally, by using the VTS module it is not possible to add custom variables to the stats, such as information about the namespace, ingress rule, and service (something available as NGINX variable).

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

replaces #2607 

fixes #2355
fixes #2128 
fixes #557

aledbf commented:
You have some lint issues (maybe from one of my commits :P )
```
./hack/verify-golint.sh
!!! 'golint' problems: 
/go/src/k8s.io/ingress-nginx/internal/ingress/metric/collector/stats.go:53:6: exported type StatsCollector should have comment or be unexported
/go/src/k8s.io/ingress-nginx/internal/ingress/metric/collector/stats.go:64:1: exported function NewInstance should have comment or be unexported
/home/aledbf/go/src/k8s.io/ingress-nginx/internal/ingress/metric/collector/stats.go:179:1: exported method StatsCollector.Run should have comment or be unexported
```

gianrubio commented:
Does it make sense to fully remove the current vts library or should we have metrics/v1 => vts and metrics/v2 => lua exporter.

Can I have a docker image of this changes?

aledbf commented:
@gianrubio I was building that :wink:  `quay.io/aledbf/nginx-ingress-controller:0.371`

aledbf commented:
> Does it make sense to fully remove the current vts library or should we have metrics/v1 => vts and metrics/v2 => lua exporter.

Once this PR lands we are removing the VTS module from the nginx image

aledbf commented:
@fmejia97 
```
I0605 19:00:21.763053       6 stats.go:126] msg: {"requestLength":"1343","namespace":"gitlab","host":"git.192.168.1.29.xip.io","upstreamResponseTime":"-","remoteAddr":"192.168.1.2","ingress":"gitlab-dev","status":"499","protocol":"HTTP\/1.1","upstreamStatus":"-","path":"\/docker-images\/base\/branches","method":"GET","service":"gitlab","requestDuration":"0.522","bytesSent":"0","upstreamIP":"10.2.1.22:80"}
panic: json: cannot unmarshal number - into Go struct field udpData.upstreamResponseTime of type float64

goroutine 66 [running]:
k8s.io/ingress-nginx/internal/ingress/metric/collector.(*StatsCollector).handleMessage(0xc4204ca5f0, 0xc420626000, 0x167, 0x10400)
	/home/aledbf/go/src/k8s.io/ingress-nginx/internal/ingress/metric/collector/stats.go:132 +0xbc7
k8s.io/ingress-nginx/internal/ingress/metric/collector.(*StatsCollector).(k8s.io/ingress-nginx/internal/ingress/metric/collector.handleMessage)-fm(0xc420626000, 0x167, 0x10400)
```

fmejia97 commented:
🤔 Looking right now

aledbf commented:
Example of the metrics https://gist.github.com/aledbf/c68ad149c1292da54d6bd3162bf05740

aledbf commented:
> "upstreamResponseTime":"-"

This could be an issue. In some scenarios, nginx returns `-` in some fields (usually when status==499)

fmejia97 commented:
Yes, that is what's crashing `Unmarshal`. I'm working on a fix and will update the PR soon.

ElvinEfendi commented:
I've not looked deep into the PR yet, but can you make another PR for Lua middleware and deletion of VTS? Will make it easier to review.

I'd suggest we first get the Lua middleware merged and then merge this PR. And finally remove VTS and edit nginx.tmpl to switch to the new metric collection.

aledbf commented:
> I'd suggest we first get the Lua middleware merged and then merge this PR. 

Both things are in this PR

> And finally remove VTS and edit nginx.tmpl to switch to the new metric collection.

That's the plan

aledbf commented:
@ElvinEfendi by default the vts module is disabled so the prometheus metrics only contain the default status information from nginx

ElvinEfendi commented:
> Both things are in this PR

I know but for example Lua middleware lacks unit tests, and having self contained separate PRs if possible makes it easier to review.

aledbf commented:
>  Will make it easier to review.

Not sure if splitting this PR makes sense (in this case). It will be harder to review from my point of view

SuperQ commented on internal/ingress/metric/collector/stats.go:
```diff
@@ -0,0 +1,181 @@
+			Name:      "bytes_sent",
+			Help:      "The the number of bytes sent to a client",
+			Namespace: ns,
+			Buckets:   prometheus.LinearBuckets(50, 25, 10), // 10 buckets, each 50 bytes wide.
```
I would probably use exponential buckets for bytes sent, from 100 bytes to maybe 10 MiB.

SuperQ commented:
Is it possible to use `ngx.socket.stream` with a unix domain socket rather than UDP? This would make more sense for something communicating locally. It would avoid the whole network stack, even on localhost large amounts of UDP can be problematic.

aledbf commented:
@SuperQ yes but not sure if that change will be done in this PR

aledbf commented:
@fmejia97 maybe we could add this additional metric https://github.com/kubernetes/ingress-nginx/pull/2121 now? (and close that PR)

fmejia97 commented:
@aledbf Looks like the PR wants to add a metric for `average request time` for individual services (upstreams). I'm exposing `upstream_response_time_seconds_sum{...}`  which contains services labels and looking at Prometheus conventions, aggregation is usually carried out at the prometheus server side. We could obtain the additional metric using the query: `avg_over_time(range-vector)` https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation-_over_time. This query allows us to specify how back in time we want to consider values for the aggregation. What are your thoughts?

andrewlouis93 commented on rootfs/etc/nginx/lua/metric_emitter.lua:
```diff
@@ -0,0 +1,56 @@
+
+function _M.call()
+  local current_phase = ngx.get_phase()
+  if current_phase == "log" then
```
Could we get rid of the `current_phase` check?

This module could potentially be called from any phase given that it just defers the actual payload emission to the `timer` phase

aledbf commented on internal/ingress/metric/collector/stats.go:
```diff
@@ -50,6 +50,7 @@ type udpData struct {
 	Service   string `json:"service"`   // Label
 }
 
+// StatsCollector: Stores prometheus metrics and ingress meta-data
```
No need of `:` in go :) `StatsCollector ...`

fmejia97 replied:
😄

fmejia97 commented:
@aledbf @andrewlouis93 @ElvinEfendi @SuperQ Addressed your comments. Also removed `metric_emitter` lua module from this PR and added it to a separate PR : https://github.com/kubernetes/ingress-nginx/pull/2611 (as requested).

gianrubio commented on internal/ingress/metric/collector/stats.go:
```diff
@@ -0,0 +1,193 @@
+	var stats udpData
+	err := json.Unmarshal(msg, &stats)
+	if err != nil {
+		panic(err)
```
This will panic only the collector or the controller?

aledbf replied:
The controller.

aledbf replied:
@fmejia97 please change that line with something like
```
glog.Errorf("Unexpected error deserializing JSON paylod: %v", err)
return
```

fmejia97 replied:
👍

gianrubio commented on internal/ingress/metric/collector/stats.go:
```diff
@@ -0,0 +1,193 @@
+	if stats.UpstreamResponseTime != -1 {
+		upstreamResponseTimeMetric, err := sc.upstreamResponseTime.GetMetricWith(labels)
+		if err != nil {
+			glog.Errorf("Error fetching upstream response time metric: %v", err)
```
Can you create a gauge metrics to show when was the last time that the collector ran without errors? You can use the same idea from [here](https://github.com/kubernetes/ingress-nginx/blob/master/internal/ingress/controller/metrics.go#L44-L57)

fmejia97 replied:
Yup :)

fmejia97 commented:
@gianrubio Added metrics to monitor the success of the collector

aledbf commented on cmd/nginx/main.go:
```diff
@@ -119,13 +119,22 @@ func main() {
 	go registerHandlers(conf.EnableProfiling, conf.ListenPorts.Health, ngx, mux)
 
-	lc, err := collector.NewInstance(conf.Namespace, class.IngressClass, 8000)
+	ngxStatusPath := "/nginx_status"
```
@fmejia97 this makes no sense now. Please remove this variable and parameter from NewNGINXStatusCollector

aledbf commented on internal/ingress/metric/collector/nginx_status_collector.go:
```diff
@@ -0,0 +1,216 @@
+	nginxStatusCollector struct {
+		scrapeChan     chan scrapeRequest
+		ngxHealthPort  int
+		ngxVtsPath     string
```
@fmejia97 please remove this field. We don't need this now.

fmejia97 commented:
@aledbf After talking with @ElvinEfendi and @andrewlouis93 , we decided that keeping `nginx_status_collector.go` is a good idea since it gives us `connections and request` metrics. Ideally we don't want to emit these metrics through the `monitor.lua` module since this module only emits data regarding each individual `request`. Connections metrics are independent of individual requests. As it is right now, the `nginx_status` collector will write metrics using the prometheus client, therefore both `monitor.lua` and `nginx_status` metrics will be visible on the same endpoing `/metrics`. If we remove it, we won't have connection metrics. What do you think?

aledbf commented:
> we decided that keeping nginx_status_collector.go is a good idea since it gives us connections and request metrics.

Until now `nginx_status_collector.go` was used only if the VTS module was disabled. I would prefer to get this info only from one place if possible. Maybe we can leave this until we merge this PR and then see exactly what can be removed (or not)

gianrubio commented:
I'd like to understand what are the main differences between the prometheus Lua collector and nginx vts? 

I noticed that nginx vts added the[ feature to expose metrics](https://github.com/vozlt/nginx-module-vts/issues/62#issuecomment-396053124) in prometheus format, can we discuss pro and con between lua x vts?

aledbf commented:
> I noticed that nginx vts added the feature to expose metrics in prometheus format, can we discuss pro and con between lua x vts?

With the dynamic mode introduced a couple of releases ago, the load balancing stuff is handled in LUA. This means there is no upstream section in the generated nginx.conf, there is only one used to switch to LUA. In this scenario, the VTS module is useless because the stats only contains information about this upstream without any context about the traffic.
Additionally to this, using the VTS module is not possible to add custom variables to the stats, like information about the namespace, ingress rule, and service (something available as NGINX variable).

andrewlouis93 commented:
@fmejia97 could you also add more context to this PR description before this ships? Is not clear to everyone why we are switching away from VTS.

aledbf commented:
@fmejia97 please rollback the last commit. The namespace cannot be a constant. Use a helper to fix the name (like ingress-nginx to ingress_nginx)

fmejia97 commented:
@aledbf I used a constant since I saw that's how it was used before. Added a new commit where I remove the constant and instead I fix the namespace name 👍

fmejia97 commented:
@aledbf Rebased + squashed 👍

andrewlouis93 commented:
@aledbf I'm happy with this as is right now. Will do another PR where we'll try an reduce complexity once this one is shipped 🙏 Great work on this @fmejia97 ❤️

aledbf commented:
Let's merge this, it's already too big.

aledbf commented:
/lgtm
/approve

aledbf commented:
@fmejia97 @andrewlouis93 I will work in the use of unix sockets instead of UDP to send the metrics
